### PR TITLE
Implement ELLA for SD1.5

### PR DESCRIPTION
**Pull request by ily-R**
This is my draft PR for the bounty **ELLA for SD1.5**. I implemented and tested the architecture. I have few steps left.

- [x]  Implement ELLA
- [x] Test the adapter's inject/eject 
- [x] implement the weight conversion script for these [weights](https://huggingface.co/QQGYLab/ELLA/discussions/5)
- [x] test ELLA contribution to SD1.5 model using **t5_xl** text_embeddings from [here](https://huggingface.co/google/flan-t5-xl)

### CODE REVIEW

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            ),
+            fl.Residual(
+                fl.Chain(
+                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                    fl.Linear(width, width * 4, device=device, dtype=dtype),
+                    SquaredReLU(),
+                    fl.Linear(width * 4, width, device=device, dtype=dtype),
+                )
+            ),
+        )
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/ella_adapter.py:157]: There are quite a few things that need to be corrected here. fl.Sum(fl.Parallel(*args)) is the same as fl.Sum(*args). This is because when the input of a sum is a tuple, it will sum it, but you can also pass multiple layers. Also  fl.Chain(fl.Sum(*args1), fl.Residual(*args2)) is the same thing that fl.Sum(*(args1 + args2)). So you can simplify everything and use a fl.Sum. It should look like this (haven't tested it):
```
class _PerceiverAttentionBlock(fl.Sum):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.GetArg(0),
            fl.Chain(
                fl.Distribute(
                    LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                ),
                fl.Parallel(
                    fl.GetArg(0),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
            ),
            fl.Chain(
                OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Linear(width, width * 4, device=device, dtype=dtype),
                SquaredReLU(),
                fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
  ```
  
Also, you should refrain from using GetArg too much. I think you should put the variable GetArg(1) in the context and remove all GetArg(0).

**Review by limiteinductive — CHANGES_REQUESTED** [review]
Thanks for the great work.

I made my first round of essentially stylistic nits. I haven't made an e2e test, but verified that conversion is working.

Could you also apply lint using `rye fmt` and verify typing with `rye run pyright`?

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 1 above)
Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:157]: You are right passing one of the inputs as a context will make the code readable and avoid using GetArg(). I ended up with the following:
`
class PerceiverAttentionBlock(fl.Chain):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.Passthrough(
                fl.UseContext("ella_perceiverresampler", "text_input"),
                InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.SetContext("ella_perceiverresampler", "proj_text_input"),
            ),
            fl.Residual(
                LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Parallel(
                    fl.Identity(),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype)
            ),
            fl.Residual(
                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    fl.Linear(width, width * 4, device=device, dtype=dtype),
                    SquaredReLU(),
                    fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
`
what do you think ?

Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:157]: the context "text_input" is set at the beginning of PerceiverResampler

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    def __init__(
+        self, width: int, output_dim: int, device: Device | str | None = None, dtype: DType | None = None
+    ) -> None:
+        super().__init__(
+            fl.Linear(width, output_dim, device=device, dtype=dtype),
+            fl.LayerNorm(output_dim, device=device, dtype=dtype),
+        )
+
+
+class PerceiverResampler(fl.Chain):
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/ella_adapter.py:187]: Out of curiosity: did you have look at the Perceiver Resampler already implemented part of IP-Adapter Plus? see https://github.com/finegrain-ai/refiners/blob/10fba3467ad3cd9e878ad3b7e797b2804cbbb261/src/refiners/foundationals/latent_diffusion/image_prompt.py#L182

Are there noticeable differences? If yes, are there ways to keep this new version consistent with the other one?

Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:187]: I did check this implementation. except for `PerceiverScaledDotProductAttention`  They are identical actually. I tried to follow the same modulation from the official ELLA repo. But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

Comment (deltheil) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:187]: > They are identical actually with few differences

OK, good to know.

In other similar cases (you might have noticed that there are several ViT-s version in Refiners - for CLIP image encoder, DINOv2, etc) we decided to have multiple versions (no DRY) precisely because there are some subtleties here and there.

> But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

I think that would be great (at least, it's worth trying!). Thanks!

Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:187]: got it, ill commit the new implementation

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 1 above)
Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:157]: following Cedric comments, ill refactor the implementation using usecontext() and following the perceiverresampler implementation on ip-adapter.

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 2 above)
Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:170]:
```
            Latents(num_latents, width, device=device, dtype=dtype),
            fl.Residual(
                fl.UseContext("ella", "timestep_embedding"),
                fl.SiLU(),
                fl.Linear(time_embedding_dim, width, device=device, dtype=dtype),
            ),
```
after checking, I found that the difference between ip perceiverresampler and ella perceiverresampler is the use of the time embeddings, they first sum the Latents with a feedforward projection of the time_embedding. They also use the time_embedding component with LayerNorm in the AdaLayerNorm module. And finally they use a squared Relu.

File: scripts/conversion/convert_ella_adapter.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert a pretrained Ella Adapter to refiners implementation")
+    parser.add_argument(
+        "--from",
+        type=str,
+        required=True,
+        dest="source_path",
+        help=("Official checkpoint from https://huggingface.co/QQGYLab/ELLA/blob/main/ella-sd1.5-tsc-t5xl.safetensors"),
+    )
```
[the remark this answers is not in the document]
Comment (limiteinductive) [reply scripts/conversion/convert_ella_adapter.py]: You can use this: https://huggingface.co/docs/huggingface_hub/guides/download

[the remark this answers is not in the document]
Comment (ily-R) [reply scripts/conversion/convert_ella_adapter.py]: done

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            ParameterInitialized(
+                num_latents,
+                width,
+                device=device,
+                dtype=dtype,
+            ),
+        )
```
Comment (Laurent2916) [inline src/refiners/foundationals/latent_diffusion/ella_adapter.py:109]: nit: I'm wondering if `ParameterInitialized` and `Latents` could be fused together in a single class. Since `Latents` is simply a chain with a single element: `ParameterInitialized`.

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        super().__init__(
+            fl.Distribute(
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+            ),
+            fl.Parallel(
+                fl.GetArg(index=1),
+                fl.Lambda(func=self.to_kv),
+                fl.Lambda(func=self.to_kv),
+            ),
```
Comment (Laurent2916) [inline src/refiners/foundationals/latent_diffusion/ella_adapter.py:130]: nit: slightly hard to read at a glance, could it be simplified ?

**Review by Laurent2916 — COMMENTED** [review]
Hey @ily-R, this looks better!
Do you see anything else @limiteinductive ?
I'll run the tests tomorrow

**Comment by Laurent2916** [comment]
I've run the tests, beside the fact that `sentencepiece` was missing from the dependencies, the 3 tests are green.

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 4 above)
Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:109]: I tried to follow the same thing you did for DINOV2 with ClassToken

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 5 above)
Comment (ily-R) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:130]: following Cedric comments, this is almost and identical implementation to IP adapter's PerceiverAttention class. Do you have a suggestion it simplify it ?

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 4 above)
Comment (Laurent2916) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:109]: ah right, we do this because this looks prettier when printing the model's repr, see https://github.com/finegrain-ai/refiners/pull/132#discussion_r1423714231

File: src/refiners/foundationals/latent_diffusion/ella_adapter.py (hunk 5 above)
Comment (Laurent2916) [reply src/refiners/foundationals/latent_diffusion/ella_adapter.py:130]: I guess this version is the less verbose one then

**Review by limiteinductive — COMMENTED** [review]
I have nothing to add (except linting and rebase); @deltheil, could you have a final look? Also, what about the `sentencepiece` dependency?

**Comment by deltheil** [comment]
> I have nothing to add (except linting and rebase);

That's great!

> @deltheil, could you have a final look?

OK, stay tuned.

> Also, what about the `sentencepiece` dependency?

@limiteinductive where is it used? ("[...] beside the fact that sentencepiece was missing from the dependencies, the 3 tests are green")

**Comment by Laurent2916** [comment]
Hi @ily-R, final nits before we merge:
- could you please add `sentencepiece` as a test dependency
- because of a recent PR, there a small conflict in `src/refiners/foundationals/latent_diffusion/stable_diffusion_1/__init__.py`, could please resolve it

**Comment by ily-R** [comment]
I rebased and did the necessary lint/format and added the sentencepiece as a test dependency . The test are green.
Let me know if there is something to fix

**Review by deltheil — COMMENTED** [review]
@ily-R sorry for the delay - I sent a few final nit-s, do you mind having a look? Also, please rebase and squash into a single commit. Thanks!

File: tests/e2e/test_diffusion.py [hunk 6]
```diff
@@ -1799,6 +1814,45 @@ def test_diffusion_textual_inversion_random_init(
...
+@no_grad()
+def test_diffusion_ella_adapter(
+    sd15_std_float16: StableDiffusion_1,
+    ella_adapter_weights: Path,
+    test_weights_path: Path,
+    expected_image_ella_adapter: Image.Image,
+    test_device: torch.device,
+):
+    sd15 = sd15_std_float16
+    t5_encoder = T5TextEmbedder(pretrained_path=test_weights_path / "QQGYLab/T5XLFP16", max_length=128).to(
```
[the remark this answers is not in the document]
Comment (ily-R) [reply tests/e2e/test_diffusion.py]: done

File: tests/e2e/test_diffusion.py [hunk 7]
```diff
@@ -506,6 +512,29 @@ def lda_ft_mse_weights(test_weights_path: Path) -> Path:
     return lda_weights
 
 
+@pytest.fixture(scope="module")
+def ella_weights(test_weights_path: Path) -> tuple[Path, Path]:
+    ella_adapter_weights = test_weights_path / "ELLA-Adapter/ella-sd1.5-tsc-t5xl.safetensors"
```
Comment (deltheil) [inline tests/e2e/test_diffusion.py:517]:
```suggestion
    ella_adapter_weights = test_weights_path / "ELLA-Adapter" / "ella-sd1.5-tsc-t5xl.safetensors"
```

Comment (deltheil) [reply tests/e2e/test_diffusion.py:517]: nit (mostly for consistency)

File: tests/e2e/test_diffusion.py [hunk 8]
```diff
@@ -590,10 +619,6 @@ def sd15_std_sde(
 def sd15_std_float16(
     text_encoder_weights: Path, lda_weights: Path, unet_weights_std: Path, test_device: torch.device
 ) -> StableDiffusion_1:
-    if test_device.type == "cpu":
-        warn("not running on CPU, skipping")
-        pytest.skip()
-
```
Comment (deltheil) [inline tests/e2e/test_diffusion.py:596]: Any reason to remove these lines? This is probably a debug leftover. Could you please undo these changes?

File: tests/adapters/test_ella_adapter.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    adapter.eject()
+    assert repr(unet) == initial_repr
+    adapter.inject()
+    assert repr(unet) != initial_repr
+    adapter.eject()
+    assert repr(unet) == initial_repr
+
+
+@no_grad()
+def test_scale(test_device: torch.device):
```
Comment (deltheil) [inline tests/adapters/test_ella_adapter.py:30]:
```suggestion
def test_ella_cross_attention(test_device: torch.device):
```

Comment (deltheil) [reply tests/adapters/test_ella_adapter.py:30]: Since there is no scale involved (it's rather testing/inspecting that the adapter properly modified the cross-attentions)

**Review by deltheil — COMMENTED** [review]
Please take a look (last round I guess)

**Comment by deltheil** [comment]
@ily-R thanks for the quick turnaround. Some last suggestions (see above).

File: tests/e2e/test_diffusion.py [hunk 10]
```diff
@@ -506,6 +512,29 @@ def lda_ft_mse_weights(test_weights_path: Path) -> Path:
     return lda_weights
 
 
+@pytest.fixture(scope="module")
+def ella_weights(test_weights_path: Path) -> tuple[Path, Path]:
+    ella_adapter_weights = test_weights_path / "ELLA-Adapter/ella-sd1.5-tsc-t5xl.safetensors"
+    if not ella_adapter_weights.is_file():
+        warn(f"could not find weights at {ella_adapter_weights}, skipping")
+        pytest.skip(allow_module_level=True)
+    t5xl_weights = test_weights_path / "QQGYLab/T5XLFP16"
```
Comment (deltheil) [inline tests/e2e/test_diffusion.py:521]:
```suggestion
    t5xl_weights = test_weights_path / "QQGYLab" / "T5XLFP16"
```

File: tests/e2e/test_diffusion.py (hunk 8 above)
Comment (ily-R) [reply tests/e2e/test_diffusion.py:596]: oh! that was a typo

**Review by limiteinductive — APPROVED** [review]
LGTM
